File: examples/mobilenetv1/python/mobilenetv1.py
#!/usr/bin/env python3
# Copyright 2026 SpacemiT (Hangzhou) Technology Co. Ltd.
#
# SPDX-License-Identifier: Apache-2.0

"""
MobileNetV1 Image Classification Example using CV Model Factory

MobileNetV1 输出 1001 类（含 background），使用 deploy.mobilenet.MobileNetV1Classifier
截掉背景类后对齐 1000 类 ImageNet 标签。

运行方式：通过 --config 指定 yaml 路径（与 yolov8.py 一致）。
"""

import sys
import argparse
from pathlib import Path
import cv2
import yaml

# Add src to path for imports
sys.path.append(str(Path(__file__).parent.parent.parent.parent / "src"))

from core.python.vision_model_factory import ModelFactory


def resolve_path(path_value, project_root):
    p = Path(path_value).expanduser()
    return p if p.is_absolute() else (project_root / p).resolve()


def load_imagenet_labels(label_file: Path) -> list:
    """ImageNet 标签文件：每行「WordNet_ID 类别名」，只取类别名。"""
    labels = []
    with open(label_file, encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            _, _, name = line.partition(" ")
            labels.append(name if name else line)
    return labels


def find_label_file(path_value: str, project_root: Path, config_dir: Path):
    p = Path(path_value).expanduser()
    if p.is_absolute():
        return p if p.is_file() else None
    for base in (project_root, config_dir.parent.parent):
        candidate = (base / p).resolve()
        if candidate.is_file():
            return candidate
    return None


def parse_args():
    parser = argparse.ArgumentParser(description="MobileNetV1 Classification Example")
    parser.add_argument(
        "--config",
        type=str,
        default=None,
        help="Config yaml path (default: examples/mobilenetv1/config/mobilenetv1.yaml)",
    )
    parser.add_argument(
        "--image",
        type=str,
        help="Input image path (if not provided, uses config default)",
    )
    parser.add_argument("--top-k", type=int, default=5, help="Show top-k predictions")
    parser.add_argument(
        "--model-path", type=str, default=None, help="Override model_path in yaml"
    )
    return parser.parse_args()


def main():
    args = parse_args()

    try:
        default_config = Path(__file__).parent.parent / "config" / "mobilenetv1.yaml"
        config_path = Path(args.config) if args.config else default_config
        config_dir = config_path.parent
        project_root = Path(__file__).parent.parent.parent.parent  # model_zoo/cv
        model_name = config_path.stem

        # Create model factory
        factory = ModelFactory()

        # Create MobileNet classifier
        print(f"创建 {model_name} 分类器...")
        override_params = {}
        if args.model_path:
            p = Path(args.model_path).expanduser()
            override_params["model_path"] = str(
                p if p.is_absolute() else (project_root / p).resolve()
            )
        classifier = factory.create_model(
            model_name,
            config_dir=config_dir,
            **override_params,
        )

        # Get model config from specified yaml (test_image, label_file_path)
        config = {}
        if config_path.exists():
            with open(config_path, "r", encoding="utf-8") as f:
                config = yaml.safe_load(f) or {}

        # Use provided image or default from config
        if args.image:
            image_path = resolve_path(args.image, project_root)
        else:
            image_path = config.get("test_image", "test_data/images/cat.jpg")
            image_path = resolve_path(image_path, project_root)

        # Load image
        print(f"加载图像: {image_path}")
        image = cv2.imread(str(image_path))
        if image is None:
            raise FileNotFoundError(f"无法加载图像: {image_path}")

        print(f"图像尺寸: {image.shape}")

        labels = []
        label_rel = config.get("label_file_path")
        if label_rel:
            label_file = find_label_file(label_rel, project_root, config_dir)
            if label_file:
                labels = load_imagenet_labels(label_file)
                print(f"加载标签文件: {label_file} ({len(labels)} 个)")
            else:
                print(f"警告: 未找到标签文件 {label_rel}")

        # Run classification（ResNetClassifier 使用 predict_top_k，返回 [(class_name, score), ...]）
        print("运行图像分类...")
        predictions = classifier.predict_top_k(image, labels or [], k=args.top_k)

        print(f"\nTop-{args.top_k} 预测结果:")
        for i, (class_name, confidence) in enumerate(predictions):
            print(f"  {i + 1}: {class_name} ({confidence:.4f})")

    except Exception as e:
        print(f"错误: {e}")
        return 1

    return 0


if __name__ == "__main__":
    exit(main())

File: examples/mobilenetv1/cpp/mobilenetv1.cpp
/*
 * Copyright (C) 2026 SpacemiT (Hangzhou) Technology Co. Ltd.
 * SPDX-License-Identifier: Apache-2.0
 */

#include <algorithm>   // NOLINT(build/include_order)
#include <fstream>     // NOLINT(build/include_order)
#include <iostream>    // NOLINT(build/include_order)
#include <memory>      // NOLINT(build/include_order)
#include <string>      // NOLINT(build/include_order)
#include <vector>      // NOLINT(build/include_order)
#include <iomanip>     // NOLINT(build/include_order)
#include <filesystem>  // NOLINT(build/c++17) NOLINT(build/include_order)

#include <opencv2/opencv.hpp>  // NOLINT(build/include_order)
#include <yaml-cpp/yaml.h>     // NOLINT(build/include_order)

#include "vision_service.h"  // NOLINT(build/include_order)

namespace {

std::string ResolveResourcePath(const std::string& path) {
    if (path.empty() || path[0] == '/' || (path.size() >= 2 && path[1] == ':')) {
        return path;
    }
    if (std::filesystem::exists(path)) return path;
    const std::string with_parent = "../" + path;
    if (std::filesystem::exists(with_parent)) return with_parent;
    return path;
}

std::string ResolveLabelFile(const std::string& path, const std::string& config_path) {
    if (path.empty()) return path;
    namespace fs = std::filesystem;
    const auto exists = [](const fs::path& p) { return !p.empty() && fs::exists(p); };
    fs::path rel(path);
    if (rel.is_absolute() && exists(rel)) return rel.string();
    if (exists(rel)) return rel.string();
    if (exists(fs::path("..") / rel)) return (fs::path("..") / rel).string();
    const fs::path config_dir = fs::path(config_path).parent_path();
    const fs::path repo_root = config_dir.parent_path().parent_path().parent_path();
    const fs::path from_repo = repo_root / rel;
    if (exists(from_repo)) return from_repo.string();
    return ResolveResourcePath(path);
}

// Load ImageNet-format labels: one entry per line, optionally prefixed by a
// WordNet ID (e.g. "n01440764 tench, Tinca tinca" -> "tench, Tinca tinca").
std::vector<std::string> LoadImagenetLabels(const std::string& label_file) {
    std::vector<std::string> labels;
    std::ifstream file(label_file);
    if (!file.is_open()) return labels;
    std::string line;
    while (std::getline(file, line)) {
        line.erase(line.find_last_not_of(" \t\r\n") + 1);
        if (line.empty()) continue;
        const size_t pos = line.find(' ');
        if (pos != std::string::npos && pos + 1 < line.size()) {
            labels.push_back(line.substr(pos + 1));
        } else {
            labels.push_back(line);
        }
    }
    return labels;
}

}  // namespace

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cout << "Usage: " << argv[0] << " <config_yaml> [options]\n"
                    << "  Options (any order): --model-path <path> --image <path>\n";
        return 1;
    }

    std::string config_path = argv[1];
    std::string image_path;
    std::string model_path_override;
    std::string label_file_path;

    for (int i = 2; i < argc; i++) {
        std::string arg = argv[i];
        if (arg == "--image" && i + 1 < argc) {
            image_path = argv[++i];
        } else if (arg == "--model-path" && i + 1 < argc) {
            model_path_override = argv[++i];
        }
    }

    if (std::filesystem::exists(config_path)) {
        try {
            YAML::Node config = YAML::LoadFile(config_path);
            if (config["label_file_path"]) {
                label_file_path = ResolveLabelFile(
                    config["label_file_path"].as<std::string>(), config_path);
            }
        } catch (...) {}
    }
    std::vector<std::string> labels = LoadImagenetLabels(label_file_path);

    std::unique_ptr<VisionService> service = VisionService::Create(
        config_path,
        model_path_override,
        true);
    if (!service) {
        std::cerr << "Error: " << VisionService::LastCreateError() << std::endl;
        return 1;
    }

    if (image_path.empty()) {
        const std::string default_image = service->GetDefaultImage();
        if (!default_image.empty()) {
            image_path = default_image;
        }
    }
    if (image_path.empty()) image_path = ResolveResourcePath("test_data/images/cat.jpg");

    std::cout << "Loading image: " << image_path << std::endl;
    cv::Mat img = cv::imread(image_path);
    if (img.empty()) {
        std::cerr << "Error: Could not load image: " << image_path << std::endl;
        return 1;
    }
    VisionServiceResponse response;
    VisionServiceStatus ret = service->Infer(img, &response);
    if (ret != VISION_SERVICE_OK) {
        std::cerr << "Error: " << service->LastError() << std::endl;
        return 1;
    }

    if (!response.results.empty()) {
        std::cout << "Top-5 classification results:" << std::endl;
        const size_t max_results = std::min<size_t>(response.results.size(), 5);
        for (size_t i = 0; i < max_results; i++) {
            int label_id = vision::get_label(response.results[i]);
            std::string name = (label_id >= 0 && static_cast<size_t>(label_id) < labels.size())
                                ? labels[static_cast<size_t>(label_id)] : ("Class " + std::to_string(label_id));
            std::cout << "  " << (i + 1) << ". " << name
                        << " (confidence: " << std::fixed << std::setprecision(4)
                        << vision::get_score(response.results[i]) << ")" << std::endl;
        }
    } else {
        std::cout << "No classification results" << std::endl;
    }
    return 0;
}
